wnfs/src/private/hamt/merge.rs

```diff
@@ -0,0 +1,214 @@
+use super::{diff, ChangeType, Node};
+use crate::{BlockStore, FsError, Hasher, Link};
+use anyhow::Result;
+use serde::de::DeserializeOwned;
+use std::{hash::Hash, rc::Rc};
+
+//--------------------------------------------------------------------------------------------------
+// Functions
+//--------------------------------------------------------------------------------------------------
+
+/// Merges a node with another with the help of a resolver function.
+pub async fn merge<K, V, H, F, B: BlockStore>(
+    main_link: Link<Rc<Node<K, V, H>>>,
+    other_link: Link<Rc<Node<K, V, H>>>,
+    f: F,
+    store: &mut B,
+) -> Result<Rc<Node<K, V, H>>>
+where
+    F: Fn(&V, &V) -> Result<V>,
+    K: DeserializeOwned + Eq + Clone + Hash + AsRef<[u8]>,
+    V: DeserializeOwned + Eq + Clone,
+    H: Hasher + Clone + 'static,
+{
+    let kv_changes = diff::kv_diff(main_link.clone(), other_link.clone(), None, store).await?;
+
+    let main_node = main_link.resolve_owned_value(store).await?;
+    let other_node = other_link.resolve_owned_value(store).await?;
+
+    let mut merge_node = Rc::clone(&main_node);
+    for change in kv_changes {
+        match change.r#type {
+            ChangeType::Remove => {
+                merge_node = merge_node
+                    .set(change.key, change.other_value.unwrap(), store)
+                    .await?;
+            }
+            ChangeType::Modify => {
+                let main_value = main_node
+                    .get(&change.key, store)
+                    .await?
+                    .ok_or(FsError::KeyNotFoundInHamt)?;
+
+                let other_value = other_node
+                    .get(&change.key, store)
+                    .await?
+                    .ok_or(FsError::KeyNotFoundInHamt)?;
+
+                merge_node = merge_node
+                    .set(change.key, f(main_value, other_value)?, store)
+                    .await?;
+            }
+            _ => (),
+        }
+    }
+
+    Ok(merge_node)
+}
+
+//--------------------------------------------------------------------------------------------------
+// Tests
+//--------------------------------------------------------------------------------------------------
+
+#[cfg(test)]
+mod proptests {
+    use crate::{
+        private::strategies::{self, operations, Operations},
+        utils::test_setup,
+        Link,
+    };
+    use async_std::task;
+    use std::{cmp, rc::Rc};
+    use test_strategy::proptest;
+
+    #[proptest(cases = 100)]
+    fn merge_associativity(
+        #[strategy(operations("[a-z0-9]{1,4}", 0u64..1000, 0..100))] ops1: Operations<String, u64>,
+        #[strategy(operations("[a-z0-9]{1,4}", 0u64..1000, 0..100))] ops2: Operations<String, u64>,
+        #[strategy(operations("[a-z0-9]{1,4}", 0u64..1000, 0..100))] ops3: Operations<String, u64>,
```

Comment:
My vote is for a slightly smaller domain for random keys & values.
This domain is ~ $2^{10}$ for the values and $36 + 36^2 + 36^3 + 36^4 \approx 2^{20}$, which gives us roughly $2^{30}$ (~1 billion) different possible key pairs. The chances that two arrays of key value pairs with length 100 chosen randomly sharing an exact key-value pair is pretty slim I think (although I didn't calculate it).
We can increase the chances either by increasing the length of the generated operations arrays or by reducing the size of keys & values.
I'd be down for a *really* small key size like `0u64..10` or even `0u64..3` (since I think in principle all that matters is whether the values are equal or not. With three HAMTs, and 0..3 you can have all of them be unequal).
And I guess reducing the size of keys generated a little helps quite a lot. (i.e. reducing it to `{1,3}` already reduces key space to just ~ $2^{15}$).